### DISCUSSION TITLE
Continue generating is not supported?

### DISCUSSION THREAD

**Discussion by zdwork** [danny-avila/LibreChat#380] [Feature Requests & Suggestions]
### What happened?

Continue generating is not supported?

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
There is no continue regenerating feature. You can simply ask the AI to continue. I will take this as a feature request

**Comment by zdwork** [comment]
Hello, when is this function expected to be launched?

![image](https://github.com/danny-avila/chatgpt-clone/assets/24469374/a64501eb-dcd4-44e0-9f0e-dffd1f72e4e6)

**Reply by zdwork** [reply]
It is more comfortable to regenerate and continue production or put it at the bottom, it is recommended

**Comment by zdwork** [comment]
It should not be possible to continue to generate, right? Is the official API also provided?

**Reply by danny-avila** [reply]
It's possible to "fake" it but the solution would involve prompt injecting and may not always result in the same output. Also, it would be hard to detect when exactly the response is cut off. I'm sure OpenAI has internal tools that make all of this easier.

**Reply by SphaeroX** [reply]
> It's possible to "fake" it but the solution would involve prompt injecting and may not always result in the same output. Also, it would be hard to detect when exactly the response is cut off. I'm sure OpenAI has internal tools that make all of this easier.

doesn't the api return how many tokens it has used? Then you could see if there are more than 2048 tokens and then show the button. In addition, you can output with each response in the UI how many tokens were used.

 **'total_tokens': 87**

```
{
 'id': 'chatcmpl-6p9XYPYSTTRi0xEviKjjilqrWU2Ve',
 'object': 'chat.completion',
 'created': 1677649420,
 'model': 'gpt-3.5-turbo',
 'usage': {'prompt_tokens': 56, 'completion_tokens': 31, 'total_tokens': 87},
 'choices': [
   {
    'message': {
      'role': 'assistant',
      'content': 'The 2020 World Series was played in Arlington, Texas at the Globe Life Field, which was the new home stadium for the Texas Rangers.'},
    'finish_reason': 'stop',
    'index': 0
   }
  ]
}
```

**Reply by zdwork** [reply]
The problem is that the official does not provide an API for continued generation

**Reply by SphaeroX** [reply]
It actually works in the playground, too. I'll have to see how they do it there as soon as I have time

**Reply by zdwork** [reply]
Come on, looking forward to it

**Reply by SphaeroX** [reply]
the first thing i did was to send a request athttps://platform.openai.com/playground?mode=chat

> {
>     "messages": [
>         {
>             "role": "user",
>             "content": "SCHREIBE MIR EINEN ARTIKEL ÜBER GEHWEGE"
>         }
>     ],
>     "temperature": 1,
>     "max_tokens": 107,
>     "top_p": 1,
>     "frequency_penalty": 0,
>     "presence_penalty": 0,
>     "model": "gpt-3.5-turbo",
>     "stream": true
> }

then came the token limit that I had set (107).

Then I sent another empty request and the playground sent the following to the chat (wich continues the respond):

> {
>     "messages": [
>         {
>             "role": "user",
>             "content": "SCHREIBE MIR EINEN ARTIKEL ÜBER GEHWEGE"
>         },
>         {
>             "role": "assistant",
>             "content": "Gehwege sind ein wichtiger Bestandteil unserer Straßen und spielen eine entscheidende Rolle für die Sicherheit von Fußgängern. Sie sind oft die einzige Möglichkeit für Fußgänger, die Straße sicher zu überqueren oder an einem Ort entlang zu gehen. Aber wie wichtig sind Gehwege in unserer Gesellschaft und wie können sie verbessert werden?\n\nWarum sind Gehwege wichtig?\n\nGehwege ermöglichen es Fußgängern, sich frei und sicher zu bewegen"
>         }
>     ],
>     "temperature": 1,
>     "max_tokens": 107,
>     "top_p": 1,
>     "frequency_penalty": 0,
>     "presence_penalty": 0,
>     "model": "gpt-3.5-turbo",
>     "stream": true
> }


then again and the following was sent:

> {
>     "messages": [
>         {
>             "role": "user",
>             "content": "SCHREIBE MIR EINEN ARTIKEL ÜBER GEHWEGE"
>         },
>         {
>             "role": "assistant",
>             "content": "Gehwege sind ein wichtiger Bestandteil unserer Straßen und spielen eine entscheidende Rolle für die Sicherheit von Fußgängern. Sie sind oft die einzige Möglichkeit für Fußgänger, die Straße sicher zu überqueren oder an einem Ort entlang zu gehen. Aber wie wichtig sind Gehwege in unserer Gesellschaft und wie können sie verbessert werden?\n\nWarum sind Gehwege wichtig?\n\nGehwege ermöglichen es Fußgängern, sich frei und sicher zu bewegen"
>         },
>         {
>             "role": "assistant",
>             "content": ". Sie sorgen dafür, dass Fußgänger von Fahrzeugen getrennt sind und somit ein höheres Maß an Sicherheit genießen können. Auf Gehwegen können Kinder zur Schule gehen, ältere Menschen spazieren und Jogger ihre Runden drehen. Sie sind auch ein wichtiger Bestandteil von städtischen Räumen, die das Leben und die Gemeinschaft fördern. Gehwege sollten deshalb so konzipiert werden, dass sie zu einer"
>         }
>     ],
>     "temperature": 1,
>     "max_tokens": 107,
>     "top_p": 1,
>     "frequency_penalty": 0,
>     "presence_penalty": 0,
>     "model": "gpt-3.5-turbo",
>     "stream": true
> }


this would be a possibility for a continue function

**Reply by zdwork** [reply]
This seems inaccurate. For example, if you give him a large code block and ask him to add comments to you and return it. When sending something to truncate, the subsequent code that continues to be generated at this time may not be submitted by you.

**Reply by danny-avila** [reply]
I saw an implementation of this somewhere and noticed 'finish_reason': 'max_tokens' is what triggers the continue generating button on the official version. Using this knowledge and @SphaeroX's example, it may be possible to send the current convo with assistant as the latest message and it will continue. Then it will be a matter of editing the frontend to handle a continuation.

This feature will officially be planned but I can't guarantee when. It's not hard but will be involved.

**Reply by zdwork** [reply]
Come on, looking forward to it, this function is very useful

**Reply by zdwork** [reply]
expect

**Comment by danny-avila** [comment]
@zdwork I'm finally getting around to this and will have it released today

similar to chatgpt, it will trigger when the api sends an irregular finish_reason

You can also manually trigger by using "stop" buttons during generation

### Here is a stopped message:
![image](https://github.com/danny-avila/LibreChat/assets/110412045/de05e732-147a-4e98-8407-81227bff2f10)



### continued:
![image](https://github.com/danny-avila/LibreChat/assets/110412045/76675fce-2b2d-49c9-942f-e77e21aa653d)

**Reply by danny-avila** [reply]
Related #658 #639

**Comment by danny-avila** [comment]
This is finally complete and wrote about it in detail in my first medium article: https://medium.com/@daniel-avila/reverse-engineering-chatgpt-the-continue-generating-function-5cb4194b0ea6

**Reply by danny-avila** [reply]
@SphaeroX I gave you a shoutout in my article!

**Comment by peakjames** [comment]
Just to flag that I'm unsure this is working as intended in all cases. 

I used the continue generating function today and it ate a _lot_ of tokens by continually repeating the already existing information rather than simply continuing from where it left off and judging by the steep use of tokens seemed to repeat the original request each time as well. 

I am wondering if it may be a quirk with it running into API minute/day limits that I'm currently throttled with (openai) and that it may operate under normal conditions.

**Reply by danny-avila** [reply]
can you give an example of what you mean? 

Sometimes when you continue it repeats. it needs a bit of coercing, it helps when you are in the middle of a sentence. if you are not, it's more likely to repeat itself.

You can always cancel generation too.

**Comment by zdwork** [comment]
Continue generating, currently not supported yet.


<img width="1162" alt="image" src="https://github.com/danny-avila/LibreChat/assets/24469374/64bbc2c4-62b3-44a0-892c-17a944f235a8">

**Reply by danny-avila** [reply]
It is supported:

<img width="252" alt="image" src="https://github.com/danny-avila/LibreChat/assets/110412045/e83c207d-5329-4c3b-9340-e82f75881851">

**Comment by zdwork** [comment]
LibreChat sometimes has the "Continue generating" button, while other times only outputs half of the content without showing the "Continue generating" button. It's strange.

**Reply by danny-avila** [reply]
You have to stop the generation for “continue” to show
